_posts/2019-07-26-pysam-with-s3.md

```diff
@@ -0,0 +1,182 @@
+---
+layout: blogpost
+title: Using PySAM to access private S3 Bucket
+author: Jackson Zheng
+date: 2019-07-26
+---
+
+I was working on a project for [CanDIG](https://www.distributedgenomics.ca/) which was an implementation of the [Htsget API](http://samtools.github.io/hts-specs/htsget.html) &mdash; a simple data-access API for reads and variants. A key component of our application was the [PySAM library](https://pysam.readthedocs.io/en/latest/), which provides an interface to genomic data. Its main use in the service was to parse a desired file and return only a chunk of that file with various filters. The most common way of accessing a file is with a local file path. However, with S3 buckets emerging as a popular file storage system, it is natural that PySAM should be able to interact with S3 objects; CanDIG, for instance, is moving to using a standard S3 interface for its data objects using [MinIO](https://min.io). PySAM is a wrapper of the htslib C-API which should support fetching of S3 objects, but many issues arose when my team and I tried to pass access objects in private S3 buckets with pysam or htslib. After many google searches, there seems to be many people having similar issues:
+1) https://www.biostars.org/p/213305/
+2) https://github.com/pysam-developers/pysam/issues/557
+3) https://www.biostars.org/p/213305/ 
+
+Although several sources claim that accessing s3 objects should just work, I have not been able to find a single worked example. After many days of digging on the web and coming across several problems, my team and I figured out how to successfully use PySAM and htslib to access private S3 bucket; we describe how below.
+
+### 1) Download and install htslib version from developer branch
+   
+We believe the root of the problem is AWS's move to a [new algorithm](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html) for signed URL access to private objects which is not supported by the current release of htslib (v1.9); but is supported on the `devel` branch.  To use this new version, the installation instructions can be found here: https://github.com/samtools/htslib . Make sure to add the --enable-s3 option during installation. 
+
+To check whether htslib installed correctly, use the command:
+```
+which htsfile
+```
+The output should be:
+```
+/usr/local/bin/htsfile
+```
+
+### 2) Open file from AWS S3 path with htslib and environment variables
+We can demonstrate that a file from a S3 path can be accessed with the new htslib. 
+
+##### Set environment variables
+```
+AWS_SECRET_ACCESS_KEY=XXXXXXXX
+AWS_ACCESS_KEY_ID=XXXXXXXX
+```
+##### Run the command
+```
+htsfile s3://1000genomes/1000G_2504_high_coverage/data/ERR3243163/NA12546.final.cram
```

Comment:
We should talk about this; unfortunately using this with a public bucket doesn't really prove anything, since it doesn't test the request-signing code path.  Let's change this so that we explain how to set up a simple test with an AWS private bucket.

Reply:
You can say this with something like: “With the devel-branch htslib, you should be able to successfully run tools like `htsfile` on a private S3 bucket, like so (using your AWS credentials in the environment variables):”